src/network/protocol/messages/merkleblock.zig

```diff
@@ -0,0 +1,146 @@
+const std = @import("std");
+const protocol = @import("../lib.zig");
+
+const Sha256 = std.crypto.hash.sha2.Sha256;
+const BlockHeader = protocol.BlockHeader;
+const CompactSizeUint = @import("bitcoin-primitives").types.CompatSizeUint;
+
+/// MerkleBlockMessage represents the "MerkleBlock" message
+///
+/// https://developer.bitcoin.org/reference/p2p_networking.html#merkleblock
+pub const MerkleBlockMessage = struct {
+    block_header: BlockHeader,
+    transaction_count: u32,
+    hash_count: CompactSizeUint,
+    hashes: []const [32]u8,
+    flag_bytes: CompactSizeUint,
+    flags: []const u8,
+
+    const Self = @This();
+
+    pub inline fn name() *const [12]u8 {
+        return protocol.CommandNames.MERKLEBLOCK ++ [_]u8{0} ** 6;
+    }
+
+    /// Returns the message checksum
+    pub fn checksum(self: *const Self) [4]u8 {
+        var digest: [32]u8 = undefined;
+        var hasher = Sha256.init(.{});
+        self.serializeToWriter(hasher.writer()) catch unreachable;
+        hasher.final(&digest);
+
+        Sha256.hash(&digest, &digest, .{});
+
+        return digest[0..4].*;
+    }
+
+    /// Free the allocated memory
+    pub fn deinit(self: *const Self, allocator: std.mem.Allocator) void {
+        allocator.free(self.flags);
+        allocator.free(self.hashes);
+    }
+
+    /// Serialize the message as bytes and write them to the Writer.
+    pub fn serializeToWriter(self: *const Self, w: anytype) !void {
+        try self.block_header.serializeToWriter(w);
+        try w.writeInt(u32, self.transaction_count, .little);
+        try self.hash_count.encodeToWriter(w);
+        for (self.hashes) |hash| {
+            try w.writeAll(&hash);
+        }
+        try self.flag_bytes.encodeToWriter(w);
+        try w.writeAll(self.flags);
+    }
+
+    /// Serialize a message as bytes and return them.
+    pub fn serialize(self: *const Self, allocator: std.mem.Allocator) ![]u8 {
+        var list = std.ArrayList(u8).init(allocator);
+        errdefer list.deinit();
+        try self.serializeToWriter(list.writer());
+        return list.toOwnedSlice();
```

Comment:
```zig
        const serialized_len = self.hintSerializedLen();

        const ret = try allocator.alloc(u8, serialized_len);
        errdefer allocator.free(ret);

        try self.serializeToSlice(ret);

        return ret;
```

no need for arraylist, you already know what size it will be